File: apps/worker/worker/evaluations/semantic.py
import hashlib
import math
from collections.abc import Mapping, Sequence


def text_hash(text: str) -> str:
	return hashlib.blake2b(text.encode(), digest_size=16).hexdigest()


def cosine_similarity(a: Sequence[float], b: Sequence[float]) -> float:
	if len(a) != len(b):
		return 0.0
	dot = sum(x * y for x, y in zip(a, b))
	norm_a = math.sqrt(sum(x * x for x in a))
	norm_b = math.sqrt(sum(y * y for y in b))
	if norm_a == 0 or norm_b == 0:
		return 0.0
	return dot / (norm_a * norm_b)


def top_semantic_matches(
	requirement_vector: Sequence[float],
	block_vectors: Mapping[str, Sequence[float]],
	*,
	limit: int = 3,
	min_similarity: float = 0.3,
) -> list[dict[str, object]]:
	similarities = (
		(block_id, round(cosine_similarity(requirement_vector, vector), 4))
		for block_id, vector in block_vectors.items()
	)
	ranked = sorted(
		[
			(block_id, similarity)
			for block_id, similarity in similarities
			if similarity >= min_similarity
		],
		key=lambda item: item[1],
		reverse=True,
	)
	return [
		{"blockId": block_id, "similarity": similarity}
		for block_id, similarity in ranked[:limit]
	]

File: apps/worker-new/src/domain/semantic.ts
import { createHash } from "node:crypto";

export const textHash = (text: string): string =>
	createHash("sha256").update(text, "utf8").digest("hex").slice(0, 32);

export const cosineSimilarity = (a: readonly number[], b: readonly number[]): number => {
	if (a.length !== b.length || !a.length) return 0;
	let dot = 0;
	let normA = 0;
	let normB = 0;
	for (let index = 0; index < a.length; index++) {
		dot += (a[index] as number) * (b[index] as number);
		normA += (a[index] as number) * (a[index] as number);
		normB += (b[index] as number) * (b[index] as number);
	}
	if (!normA || !normB) return 0;
	return dot / (Math.sqrt(normA) * Math.sqrt(normB));
};

export const topSemanticMatches = (
	requirementVector: readonly number[],
	blockVectors: Readonly<Record<string, readonly number[]>>,
	options: { limit?: number; minSimilarity?: number } = {},
): Array<{ blockId: string; similarity: number }> => {
	const limit = options.limit ?? 3;
	const minSimilarity = options.minSimilarity ?? 0.3;
	return Object.entries(blockVectors)
		.map(([blockId, vector]) => ({ blockId, similarity: Math.round(cosineSimilarity(requirementVector, vector) * 10000) / 10000 }))
		.filter((item) => item.similarity >= minSimilarity)
		.sort((left, right) => right.similarity - left.similarity)
		.slice(0, limit);
};
